src/DeploymentFile.ts

```diff
@@ -0,0 +1,135 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import { CachedValue } from "./CachedValue";
+import { INamedDefinition } from "./INamedDefinition";
+import * as Json from "./JSON";
+import * as language from "./Language";
+import { ReferenceList } from "./ReferenceList";
+import { nonNullOrEmptyValue } from "./util/nonNull";
+
+/**
+ * Represents a deployment-related JSON file
+ */
+export abstract class DeploymentFile {
+    // Parse result for the template JSON document as a whole
+    private _jsonParseResult: Json.ParseResult;
+
+    // The JSON node for the top-level JSON object (if the JSON is not empty or malformed)
+    private _topLevelValue: Json.ObjectValue | undefined;
+
+    private _schema: CachedValue<Json.StringValue | undefined> = new CachedValue<Json.StringValue | undefined>();
+
+    /**
+     * Constructor
+     *
+     * @param _documentText The string text of the document.
+     * @param _documentId A unique identifier for this document. Usually this will be a URI to the document.
+     */
+    constructor(private _documentText: string, private _documentId: string) {
+        nonNullOrEmptyValue(_documentId, "_documentId");
+
+        this._jsonParseResult = Json.parse(_documentText);
+        this._topLevelValue = Json.asObjectValue(this._jsonParseResult.value);
+    }
+
+    /**
+     * Get the document text as a string.
+     */
+    public get documentText(): string {
+        return this._documentText;
+    }
+
+    /**
+     * The unique identifier for this deployment template. Usually this will be a URI to the document.
+     */
+    public get documentId(): string {
+        return this._documentId;
+    }
+
+    // Parse result for the template JSON document as a whole
+    public get jsonParseResult(): Json.ParseResult {
+        return this._jsonParseResult;
+    }
+
+    // The JSON node for the top-level JSON object (if the JSON is not empty or malformed)
+    public get topLevelValue(): Json.ObjectValue | undefined {
+        return this._topLevelValue;
+    }
+
+    public get schemaUri(): string | undefined {
+        const schema = this.schemaValue;
+        return schema ? schema.unquotedValue : undefined;
+    }
+
+    public get schemaValue(): Json.StringValue | undefined {
+        return this._schema.getOrCacheValue(() => {
+            const value: Json.ObjectValue | undefined = Json.asObjectValue(this._jsonParseResult.value);
```

Comment:
>const value: Json.ObjectValue | undefined = Json.asObjectValue(this._jsonParseResult.value); [](start = 12, length = 92)

Isn't this the same as _topLevelValue?

Reply:
Yep, thanks.

Reply:
return this._schema.getOrCacheValue(() => {
            return Json.asStringValue(this.topLevelValue?.getPropertyValue("$schema"));
        });